drivers/adc/ad469x/ad469x.c

```diff
@@ -0,0 +1,524 @@
+ *        Enter register mode to read/write registers
+ * @param [in] dev - ad469x_dev device handler.
+ * @param [in] channel - ad469x selected channel.
+ * @param [out] buf - data buffer.
+ * @param [in] samples - sample number.
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
+ */
+int32_t ad469x_read_data(struct ad469x_dev *dev,
+			 uint8_t channel,
+			 uint32_t *buf,
+			 uint16_t samples)
+{
+	int32_t ret;
+	uint32_t commands_data[1];
+	struct spi_engine_offload_message msg;
+	uint32_t spi_eng_msg_cmds[3] = {
+		CS_LOW,
+		WRITE_READ(1),
+		CS_HIGH
+	};
+
+	commands_data[0] = AD469x_CMD_CONFIG_CH_SEL(channel) << 8; /* ch 0 */
+
+	pwm_enable(dev->trigger_pwm_desc);
+
+	ret = spi_engine_offload_init(dev->spi_desc, dev->offload_init_param);
+	if (ret != SUCCESS)
+		return FAILURE;
+
+	msg.commands = spi_eng_msg_cmds;
+	msg.no_commands = ARRAY_SIZE(spi_eng_msg_cmds);
+	msg.rx_addr = (uint32_t)buf;
+	msg.commands_data = commands_data;
+
+	ret = spi_engine_offload_transfer(dev->spi_desc, msg, samples * 2);
+	if (ret != SUCCESS)
+		return ret;
+
+	Xil_DCacheInvalidateRange(msg.rx_addr, samples * 4);
+
+	return ret;
+}
+
+/**
+ * Initialize the device.
+ * @param [out] device - The device structure.
+ * @param [in] init_param - The structure that contains the device initial
+ * 		parameters.
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
+ */
+int32_t ad469x_init(struct ad469x_dev **device,
+		    struct ad469x_init_param *init_param)
+{
+	struct ad469x_dev *dev;
+	int32_t ret;
+	uint8_t data = 0;
+
+	dev = (struct ad469x_dev *)malloc(sizeof(*dev));
+	if (!dev)
+		return FAILURE;
+
+	ret = axi_clkgen_init(&dev->clkgen, init_param->clkgen_init);
+	if (ret != SUCCESS) {
+		printf("error: %s: axi_clkgen_init() failed\n",
+		       init_param->clkgen_init->name);
+		goto error_dev;
+	}
+
+	ret = axi_clkgen_set_rate(dev->clkgen, init_param->axi_clkgen_rate);
+	if (ret != SUCCESS) {
+		printf("error: %s: axi_clkgen_set_rate() failed\n",
+		       init_param->clkgen_init->name);
+		goto error_clkgen;
+	}
+
+	ret = ad469x_init_gpio(dev, init_param);
+	if (ret != SUCCESS)
+		goto error_clkgen;
+
+	ret = spi_init(&dev->spi_desc, init_param->spi_init);
+	if (ret != SUCCESS)
+		goto error_gpio;
+
+	dev->offload_init_param = init_param->offload_init_param;
+
+	dev->reg_access_speed = init_param->reg_access_speed;
+	dev->reg_data_width = init_param->reg_data_width;
+	dev->capture_data_width = init_param->capture_data_width;
+	dev->dev_id = init_param->dev_id;
+
+	ret = ad469x_spi_reg_write(dev, AD469x_REG_SCRATCH_PAD, AD469x_TEST_DATA);
+	if (ret < 0)
+		goto error_spi;
+
+	ad469x_spi_reg_read(dev, AD469x_REG_SCRATCH_PAD, &data);
+	if (ret != SUCCESS)
+		goto error_spi;
+
+	if (data != AD469x_TEST_DATA)
+		goto error_spi;
+
+	ret = ad469x_set_busy(dev, AD469x_busy_gp0);
+	if (ret != SUCCESS)
+		goto error_spi;
+
+	ret = ad469x_set_channel_sequence(dev, AD469x_single_cycle);
+	ad469x_enter_conversion_mode(dev);
```

Comment:
check ret before entering conversion mode. Maybe the channels sequence failed so there is not need to enter this mode before checking.